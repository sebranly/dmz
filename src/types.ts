// TODO: remove
export enum Template {
  Typescript = 'typescript',
  Javascript = 'javascript'
}

export type TemplateLanguage = {
  language: Template;
};
