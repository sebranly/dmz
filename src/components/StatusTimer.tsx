import classnames from 'classnames';
import * as React from 'react';
import { APITimer, APITimerData, TimerFrequency, TimeUnit } from '../types';
import { getNextStatus, getNextTime, getWeeklyTime } from '../utils';
import { convertSecondsToTimeValue } from '../utils/convert';
import { displayWithTwoDigits, getTimeUnitAbbreviation } from '../utils/display';
import { getSafeColor, getTimerClasses } from '../utils/tailwind';

export interface StatusTimerProps {
  className?: string;
  currentTimestamp: number;
  timer: APITimer;
}

const StatusTimer: React.FC<StatusTimerProps> = (props) => {
  // TODO: invert colors for text status
  const { className, currentTimestamp, timer } = props;

  const { data, frequency, subtitle, title } = timer;

  if (!data || data.length < 2) return null;

  const nextStatusTime = getNextStatus(currentTimestamp, timer);

  if (!nextStatusTime) return null;

  const { color: colorUnsafe, time: statusTime, textOverride } = nextStatusTime;
  const color = getSafeColor(colorUnsafe);

  const nextTime = getNextTime(currentTimestamp, statusTime, frequency);
  const remainingSeconds = nextTime - currentTimestamp;

  const statusTitle = textOverride?.title ?? title;
  const statusSubtitle = textOverride?.subtitle ?? subtitle;

  const classnamesStatusColor = `text-${color}-500`;
  const classnamesSubtitle = 'font-bold my-1 text-lg';
  const classnamesTitle = classnames(classnamesSubtitle, classnamesStatusColor);

  const {
    [TimeUnit.Day]: days,
    [TimeUnit.Hour]: hours,
    [TimeUnit.Minute]: minutes,
    [TimeUnit.Second]: seconds
  } = convertSecondsToTimeValue(remainingSeconds);

  const isFixedDays = days === 0;
  const isFixedHours = isFixedDays && hours === 0;
  const isFixedMinutes = isFixedHours && minutes === 0;
  const isFixedSeconds = isFixedMinutes && seconds === 0;

  const classnamesComponent = getTimerClasses(color, className);

  const items = [
    { value: days, label: TimeUnit.Day },
    { value: hours, label: TimeUnit.Hour },
    { value: minutes, label: TimeUnit.Minute },
    { value: seconds, label: TimeUnit.Second }
  ].map((element: { value: number; label: TimeUnit }) => {
    const { value, label: l } = element;
    const isFixed =
      (l === TimeUnit.Day && isFixedDays) ||
      (l === TimeUnit.Hour && isFixedHours) ||
      (l === TimeUnit.Minute && isFixedMinutes) ||
      (l === TimeUnit.Second && isFixedSeconds);

    const classnamesValue = classnames('font-bold text-4xl', { [classnamesStatusColor]: isFixed });

    return (
      <li className="inline-block mx-2.5" key={l}>
        <span className={classnamesValue}>{displayWithTwoDigits(value)}</span>
        <div className="text-sm">{getTimeUnitAbbreviation(l)}</div>
      </li>
    );
  });

  return (
    <div className={classnamesComponent}>
      <div className={classnamesTitle}>{statusTitle}</div>
      {statusSubtitle && <div className={classnamesSubtitle}>{statusSubtitle}</div>}
      <ul className="timer-card flex justify-center">{items}</ul>
      {/** TODO: Why is weekly forced here? */}
      {frequency === TimerFrequency.Weekly && (
        <div className="text-xs sm:text-sm">
          {data.map((dataEl: APITimerData) => {
            const { color: colorUnsafe, description, time } = dataEl;
            const color = getSafeColor(colorUnsafe);

            const classnamesTime = classnames('font-bold pr-1', `text-${color}-500`);
            return description ? (
              <div className="flex text-left pl-2.5" key={dataEl.time}>
                <div className="grow">{description}</div> <div className={classnamesTime}>{getWeeklyTime(time)}</div>
              </div>
            ) : null;
          })}
        </div>
      )}
    </div>
  );
};

export { StatusTimer };
