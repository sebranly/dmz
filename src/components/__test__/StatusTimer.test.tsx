import * as React from 'react';
import { render } from '@testing-library/react';
import { StatusTimer } from '../StatusTimer';
import { APITime, TimeFrequency, TimeStatus, TimeType } from '../../types';

test('StatusTimer', () => {
  const times: APITime[] = [
    {
      type: TimeType.Status,
      title: 'Building 21',
      frequency: TimeFrequency.Weekly,
      data: [{time: 1678471200, status: TimeStatus.Opening,}]
    },
    {
      type: TimeType.Status,
      title: 'Building 21',
      frequency: TimeFrequency.Weekly,
      data: [{time: 1678125600, status: TimeStatus.Closing,}]
    }
  ];

  const createComponent = (currentTimestamp: number) => (
    <StatusTimer currentTimestamp={currentTimestamp} time={times} />
  );

  const { container: containerOpen } = render(createComponent(1678471201));
  expect(containerOpen.childNodes).toMatchSnapshot();

  const { container: containerClosed } = render(createComponent(1678125601));
  expect(containerClosed.childNodes).toMatchSnapshot();
});
