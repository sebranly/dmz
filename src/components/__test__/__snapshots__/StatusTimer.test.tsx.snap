// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusTimer 1`] = `
NodeList [
  <div
    class="bg-neutral-800 border-solid border-2 rounded-lg border-orange-500 m-2.5 p-2.5 relative w-72 sm:w-80"
  >
    <div
      class="font-bold my-1 text-lg text-orange-500"
    >
      Building 21 is open
    </div>
    <div
      class="font-bold my-1 text-lg"
    >
      It closes in
    </div>
    <ul
      class="timer-card flex justify-center"
    >
      <li
        class="inline-block mx-2.5"
      >
        <span
          class="font-bold text-4xl"
        >
          02
        </span>
        <div
          class="text-sm"
        >
          days
        </div>
      </li>
      <li
        class="inline-block mx-2.5"
      >
        <span
          class="font-bold text-4xl"
        >
          22
        </span>
        <div
          class="text-sm"
        >
          hrs
        </div>
      </li>
      <li
        class="inline-block mx-2.5"
      >
        <span
          class="font-bold text-4xl"
        >
          59
        </span>
        <div
          class="text-sm"
        >
          min
        </div>
      </li>
      <li
        class="inline-block mx-2.5"
      >
        <span
          class="font-bold text-4xl"
        >
          59
        </span>
        <div
          class="text-sm"
        >
          sec
        </div>
      </li>
    </ul>
    <div
      class="text-xs sm:text-sm"
    >
      <div
        class="flex text-left pl-2.5"
      >
        <div
          class="grow"
        >
          Weekly Opening:
        </div>
         
        <div
          class="font-bold pr-1 text-green-500"
        >
          Fri 06:00 PM
        </div>
      </div>
      <div
        class="flex text-left pl-2.5"
      >
        <div
          class="grow"
        >
          Weekly Closing:
        </div>
         
        <div
          class="font-bold pr-1 text-orange-500"
        >
          Mon 06:00 PM
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`StatusTimer 2`] = `
NodeList [
  <div
    class="bg-neutral-800 border-solid border-2 rounded-lg border-green-500 m-2.5 p-2.5 relative w-72 sm:w-80"
  >
    <div
      class="font-bold my-1 text-lg text-green-500"
    >
      Building 21 is closed
    </div>
    <div
      class="font-bold my-1 text-lg"
    >
      It opens in
    </div>
    <ul
      class="timer-card flex justify-center"
    >
      <li
        class="inline-block mx-2.5"
      >
        <span
          class="font-bold text-4xl"
        >
          03
        </span>
        <div
          class="text-sm"
        >
          days
        </div>
      </li>
      <li
        class="inline-block mx-2.5"
      >
        <span
          class="font-bold text-4xl"
        >
          22
        </span>
        <div
          class="text-sm"
        >
          hrs
        </div>
      </li>
      <li
        class="inline-block mx-2.5"
      >
        <span
          class="font-bold text-4xl"
        >
          59
        </span>
        <div
          class="text-sm"
        >
          min
        </div>
      </li>
      <li
        class="inline-block mx-2.5"
      >
        <span
          class="font-bold text-4xl"
        >
          59
        </span>
        <div
          class="text-sm"
        >
          sec
        </div>
      </li>
    </ul>
    <div
      class="text-xs sm:text-sm"
    >
      <div
        class="flex text-left pl-2.5"
      >
        <div
          class="grow"
        >
          Weekly Opening:
        </div>
         
        <div
          class="font-bold pr-1 text-green-500"
        >
          Fri 06:00 PM
        </div>
      </div>
      <div
        class="flex text-left pl-2.5"
      >
        <div
          class="grow"
        >
          Weekly Closing:
        </div>
         
        <div
          class="font-bold pr-1 text-orange-500"
        >
          Mon 06:00 PM
        </div>
      </div>
    </div>
  </div>,
]
`;
